### PR TITLE
async-2: ChunkLoader Thread Pool

### PR DESCRIPTION
**Pull request by pwinston**
# Description
* New `NAPARI_ASYNC` env var to toggle on async loading via a thread pool.
* See more async PRs: [The Napari Rendering GitHub Project](https://github.com/users/pwinston/projects/1)

### CODE REVIEW

**Comment by pwinston** [comment]
This is ready to review. This is a smaller step that outlined in #1510 but I think it's needed. I left out the Cache and all the statistics which seemed like they could go in as separate steps.

**Comment by pwinston** [comment]
Still ready for review. I just did some cleanup and tried to anticipate some objections.

**Review by sofroniewn — APPROVED** [review]
This looks good to me @pwinston I have given the functionality a spin and can confirm the non-asyn behaviour is still as expected.

For async the basics are ok, but I did notice a couple of things are off like there are no thumbnails, and the image doesn't go to black/ when loading and the multiscale doesn't work (camera jumps around a lot). See this gif of the async working where the black and white image comes in later as it is a processed version of the magma image.

async-2_a

I think this behaviour is fine, whole point is to role out the full functionality slowly, but I just wanted to document this in case others try this out.

I've read through the code and it looks good to me, I don't have additional comments, but maybe @jni can do a deep review before merge.

File: napari/_qt/qt_chunk_receiver.py
Comment (jni) [inline napari/_qt/qt_chunk_receiver.py]: Sorry can you explain this? I'm lost now with the multiple event nesting...

File: napari/_qt/widgets/qt_layerlist.py
Comment (jni) [inline napari/_qt/widgets/qt_layerlist.py:72]: Is this doing anything? Or is it just that it needs to live *somewhere*?

File: napari/components/chunk/_info.py
Comment (jni) [inline napari/components/chunk/_info.py:11]: I suspect that you'll want to keep this until the end of the async train but I feel like we can put this class on the chopping block also!

File: napari/components/chunk/_loader.py
Comment (jni) [inline napari/components/chunk/_loader.py:189]:
```suggestion
                f'ChunkLoader.clear_pending: {num_cleared} of {num_before} '
                f'cleared -> {num_after} remain'
```

Not positive this is better, but it feels more like modern Python. The C-style % formatting is increasingly rare.

File: napari/components/chunk/_loader.py
Comment (jni) [inline napari/components/chunk/_loader.py:148]:
```suggestion
        LOGGER.debug(f'ChunkLoader._load_async: {request.key}')
```

File: napari/components/chunk/_loader.py
Comment (jni) [inline napari/components/chunk/_loader.py:238]:
```suggestion
        LOGGER.debug(f'ChunkLoader._done: {request.key}')
```

File: napari/components/chunk/_loader.py
Comment (jni) [inline napari/components/chunk/_loader.py]: Generally I like importing the module and accessing attributes with `.`, but in this case I think it clutters a lot of the code (`self.futures` (the list of futures) vs `futures` (the module), `futures.Future` in type signatures). So, a weak suggestion (totally optional) to change this to

```python
from concurrent.futures import Future, ThreadPoolExecutor, CancelledError
```

and make the corresponding changes in the code.

File: napari/components/chunk/_loader.py
Comment (jni) [inline napari/components/chunk/_loader.py]: Just pure curiosity, why is this a staticmethod?

File: napari/components/chunk/_request.py
Comment (jni) [inline napari/components/chunk/_request.py:91]: This is a great little hack. =)

File: napari/components/chunk/_utils.py
Comment (jni) [inline napari/components/chunk/_utils.py]:
```suggestion
    return id(data)  # Not a list, just use it.
```

File: napari/_qt/qt_chunk_receiver.py
Comment (jni) [inline napari/_qt/qt_chunk_receiver.py:130]:
```suggestion
            'QtChunkReceiver._on_chunk_loaded_gui: '
            f'data_id={request.key.data_id}'
```

**Review by jni — COMMENTED** [review]
Thanks @pwinston!

I've made mostly small style changes/typo fixes. The LayerInfo class seems lightweight enough to remove but you might want to leave that to later PRs. Finally, the biggest things for clarification for me are:

- I'm unclear about the whole QtChunkReceiver/QtGuiEvent machinery... Can you explain that? I didn't follow the earlier discussion in the Zulip chat, and this part is really testing my understanding of threads and Qt.
- I'm also unclear about the event nesting, `event.event.layer`... Perhaps there is a way to rename some attributes that will make this code clearer?

Other than that, onwards and upwards! Great that this already adds some async functionality — didn't test it locally but @sofroniewn's little demo speaks volumes. =D

btw @sofroniewn your comments about multiscale not working properly are *only* for NAPARI_ASYNC=1, right? It might actually be worth exploring why that's happening, since I don't see anything in the code here that would look suspicious in that respect...!

**Comment by jni** [comment]
> tried to anticipate some objections.

Nobody expects the Spanish Inquisition!!!

(Couldn't resist, I mean, it's friggin Python code...)

File: napari/_qt/widgets/qt_layerlist.py
Comment (pwinston) [reply napari/_qt/widgets/qt_layerlist.py:72]: It needs to live somewhere and it's doing something, it's moving chunks from the worker thread to the GUI thread.

File: napari/_qt/qt_chunk_receiver.py
Comment (pwinston) [reply napari/_qt/qt_chunk_receiver.py]: I added two big comments to that file and renamed `event` to `original_event`.

File: napari/components/chunk/_info.py
Comment (pwinston) [reply napari/components/chunk/_info.py:11]: LayerInfo gets really big as you can see in #1510. It ends up storing all the statistics that we use to do the auto-async stuff.

File: napari/components/chunk/_loader.py
Comment (pwinston) [reply napari/components/chunk/_loader.py:189]: With f-strings it will format all the strings, then call into the logging module, then perhaps 99.999% of the time it will discover the log-level is such that the logging is disabled, so the formatting was wasted effort.

This is why the logging module uses the older style of string formatting. Pylint wil flag you if you try to use other types of string formatting, which is the only reason I know about this guideline, from projects in the past that used pylint. Logging performance when logging is disabled is important, since that's the most common case, it will mostly be disabled.

See also:
https://docs.python.org/3/howto/logging.html#optimization
https://blog.pilosus.org/posts/2020/01/24/python-f-strings-in-logging/

There is potentially a way to do it in the "Use of alternative formatting styles" section of the logging cookbook, but it involves creating a new custom class for every log message:
https://docs.python.org/3/howto/logging-cookbook.html

I'd like to switch the async logging to JSON format which would wipe away this issue as well, although it might create other performance issues. Ultimately once in a while it would be good to time things with all the log messages totally commented out, and see if it's slowing things down even when disabled.

File: napari/components/chunk/_loader.py
Comment (pwinston) [reply napari/components/chunk/_loader.py:148]: Same comment as above about logging performance when disabled.

File: napari/components/chunk/_loader.py
Comment (pwinston) [reply napari/components/chunk/_loader.py:238]: Same comment as above about logging performance when disabled.

File: napari/components/chunk/_loader.py
Comment (pwinston) [reply napari/components/chunk/_loader.py]: Okay changed this.

File: napari/components/chunk/_loader.py
Comment (pwinston) [reply napari/components/chunk/_loader.py]: It's a staticmethod because it doesn't use self. Pylint will warn if a method doesn't use self. You can either make it a staticmethod or make it a stand-alone function. I usually make it a stand-alone function if it seems like something generic that we might use more than once. And I tend to make it a staticmethod if it seems very tied to class, something we aren't going to use again. Here I really could go either way, it seems like a toss-up.

But napari style could offer guidance, when to make a staticmethod (never, always, sometimes?) vs. a stand-alone function. If we wanted to try to be consistent. Either way it will execute identically, just has to live somewhere.

File: napari/components/chunk/_request.py
Comment (pwinston) [reply napari/components/chunk/_request.py:91]: I only changed that after the splitting. It was a regular loop before.

File: napari/_qt/qt_chunk_receiver.py
Comment (pwinston) [reply napari/_qt/qt_chunk_receiver.py:130]: Same comment as above about logging performance when disabled.

File: napari/components/chunk/_utils.py
Comment (pwinston) [reply napari/components/chunk/_utils.py]: Fixed.

File: napari/components/chunk/_loader.py
Comment (jni) [reply napari/components/chunk/_loader.py:189]: Fascinating! Thank you for the explanation. 👍 Perhaps put those two links in a comment above the first usage? That way I won't be tempted to "fix" it in 6 months when I come across this code and have forgotten this conversation. 😅

**Review by jni — APPROVED** [review]

File: napari/components/chunk/_loader.py
Comment (jni) [reply napari/components/chunk/_loader.py]: In the future I generally would prefer functions, but definitely not a big deal here. As you say — one for the style guide.

File: napari/components/chunk/_loader.py
Comment (pwinston) [reply napari/components/chunk/_loader.py:189]: I added a comment about the string formatting.

**Comment by pwinston** [comment]
The visibility and thumbnails gets added in `async-3`, see table now at the description at the top of this PR.

**Pull request merged**
